src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1121,81 +1022,110 @@ void CASESession::SendErrorMsg(SigmaErrorType errorCode)
                    ChipLogError(SecureChannel, "Failed to send error message"));
 }
 
-CHIP_ERROR CASESession::FindValidTrustedRoot(const System::PacketBufferTLVReader & tlvReader, uint32_t nTrustedRoots)
+CHIP_ERROR CASESession::GenerateDestinationID(const ByteSpan & random, const P256PublicKeySpan & rootPubkey, NodeId nodeId,
+                                              FabricId fabricId, const ByteSpan & ipk, MutableByteSpan & destinationId)
 {
-    CertificateKeyId trustedRoot;
-    System::PacketBufferTLVReader suppTlvReader;
-    uint8_t trustedRootId[kTrustedRootIdSize];
+    HMAC_sha hmac;
+    uint8_t destinationMessage[kDestinationMessageLen];
+
+    Encoding::LittleEndian::BufferWriter bbuf(destinationMessage, sizeof(destinationMessage));
+
+    bbuf.Put(random.data(), random.size());
+    bbuf.Put(rootPubkey.data(), rootPubkey.size());
+    bbuf.Put64(fabricId);
+    bbuf.Put64(nodeId);
+
+    VerifyOrReturnError(bbuf.Fit(), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    ReturnErrorOnFailure(hmac.HMAC_SHA256(ipk.data(), ipk.size(), destinationMessage, sizeof(destinationMessage),
+                                          destinationId.data(), destinationId.size()));
+
+    return CHIP_NO_ERROR;
+}
 
-    trustedRoot = CertificateKeyId(trustedRootId);
+CHIP_ERROR CASESession::FindDestinationIdCandidate(const ByteSpan & destinationId, const ByteSpan & initiatorRandom,
+                                                   const ByteSpan * ipkList, size_t ipkListEntries)
+{
+    uint8_t nCertificateSets = mOpCredSet->GetCertCount();
 
-    for (uint32_t i = 0; i < nTrustedRoots; ++i)
+    for (uint8_t i = 0; i < nCertificateSets; ++i)
```

Comment:
Suggest `size_t` and replace `i` with `certChainIdx` or such.

Reply:
Updated